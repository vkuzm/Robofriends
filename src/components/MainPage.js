import React from 'react';
import Header from '../components/Header';
import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundry from '../components/ErrorBoundry'
import '../components/mainPage.css';

class MainPage extends React.Component {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = () => {
    return this.props.robots.filter(robot => {
      return robot.name.toLowerCase().includes(this.props.searchField.toLowerCase());
    });
  }

  render() {
    const { onSearchChange, isPending, robots, error } = this.props;

    if (isPending) {
      return <div className='tc pa5 ma5'><h1>Loading...</h1></div>

    } else if (error) {
      return <div className='tc pa5 ma5'><h1>{error.message}</h1></div>

    } else {
      return (
        <div className='tc'>
          <Header />
          <SearchBox searchChange={onSearchChange} />
          <Scroll>
            <ErrorBoundry>
                <CardList robots={this.filterRobots()} />
            </ErrorBoundry>
          </Scroll>
        </div>
      );
    }
  }
}

export default MainPage;