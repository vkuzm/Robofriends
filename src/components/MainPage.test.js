import React from 'react';
import MainPage from './MainPage';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const mockGlobalProps = {
	onRequestRobots: jest.fn(),
	robots: [],
	searchField: '',
	isPending: false,
	error: null
}


it('expect to render MainPage component', () => {
	const wrapper = shallow(<MainPage {...mockGlobalProps} />);
	expect(toJson(wrapper)).toMatchSnapshot();
});


it('should shows error message when error has occurred', () => {
	const errorMessage = 'Error message';
	const mockProps = Object.assign({}, mockGlobalProps, { error: {message: errorMessage} });
	const wrapper = shallow(<MainPage {...mockProps} />);

	expect(toJson(wrapper)).toMatchSnapshot();
	expect(wrapper.find('[className="tc pa5 ma5"]').text()).toEqual(errorMessage);
});


it('should shows loading message when status is pending', () => {
	const pendingMessage = 'Loading...';
	const mockProps = Object.assign({}, mockGlobalProps, { isPending: true });
	const wrapper = shallow(<MainPage {...mockProps} />);

	expect(toJson(wrapper)).toMatchSnapshot();
	expect(wrapper.find('[className="tc pa5 ma5"]').text()).toEqual(pendingMessage);
});


it('should filters robots correctly', () => {
	const mockProps = Object.assign({}, mockGlobalProps, {
		robots: [{
			id: 1,
			name: 'Max',
			email: 'max@mail.com'
		},
		{
			id: 2,
			name: 'Martin',
			email: 'martin@mail.com'
		}],
		searchField: 'm',
	});

	const wrapper = shallow(<MainPage {...mockProps} />);
	expect(wrapper.instance().filterRobots()).toEqual(mockProps.robots);
});

it('should filters robots and returns empty array when does not find anything', () => {
	const mockProps = Object.assign({}, mockGlobalProps, {
		robots: [{
			id: 1,
			name: 'Max',
			email: 'max@mail.com'
		}],
		searchField: 'f',
	});

	const wrapper = shallow(<MainPage {...mockProps} />);
	expect(wrapper.instance().filterRobots()).toEqual([]);
});