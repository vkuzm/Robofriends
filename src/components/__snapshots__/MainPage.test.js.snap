// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render MainPage component 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox />
  <Scroll>
    <ErrorBoundry>
      <CardList
        robots={Array []}
      />
    </ErrorBoundry>
  </Scroll>
</div>
`;

exports[`should shows error message when error has occurred 1`] = `
<div
  className="tc pa5 ma5"
>
  <h1>
    Error message
  </h1>
</div>
`;

exports[`should shows loading message when status is pending 1`] = `
<div
  className="tc pa5 ma5"
>
  <h1>
    Loading...
  </h1>
</div>
`;
